### ISSUE TITLE
slic is much slower when using a mask

### ISSUE THREAD

**Issue by emmanuelle** [scikit-image/scikit-image#4902]
```python
from skimage import segmentation, data
img = data.astronaut()
mask = np.ones(img.shape[:2], dtype=np.bool)
mask[0, 0] = False
%timeit _ = segmentation.slic(img, start_label=1)
```
Output
```
224 ms ± 2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
But
```python
%timeit _ = segmentation.slic(img, mask=mask, start_label=1)
```
Output
```
1.22 s ± 23.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
The difference is very large! I wonder if it's possible to reduce it.

**Comment by grlee77** [comment]
I looked at #3120 earlier today and did not see an immediate way to speed up the Cython code, but did end up speeding up coming up with a way to accelerate the `_get_mask_centroids` a lot. I will open a PR with what I have to discuss.

**Comment by grlee77** [comment]
I came across the following repository that has a fast, MIT-licensed Cython implementation, but I don't think it supports using a mask.
https://github.com/Algy/fast-slic

It may be worth looking into that and seeing what tricks lead to the claimed performance: "Fast-slic can process 1280x720 image stream at 60fps"

**Comment by jni** [comment]
@grlee77 this issue was solved by #4903 (thank you!). Should we close it and create a separate one for the fast-slic audit?

**Comment by grlee77** [comment]
> @grlee77 this issue was solved by #4903 (thank you!). Should we close it and create a separate one for the fast-slic audit?

sure I will open a new issue
